File: lua/falcon_core/math/symbolunit.lua
-- symbolunit.lua
local cdef = require("falcon_core.ffi.cdef")
local lib = cdef.lib
local song = require("falcon_core.utils.song")
local SymbolUnit = {}

function SymbolUnit.from_string(unit_str)
    local creators = {
        ["V"] = lib.SymbolUnit_create_volt,
        ["mV"] = lib.SymbolUnit_create_millivolt,
        ["kV"] = lib.SymbolUnit_create_kilovolt,
        ["A"] = lib.SymbolUnit_create_ampere,
        ["mA"] = lib.SymbolUnit_create_milliampere,
        ["uA"] = lib.SymbolUnit_create_microampere,
        ["nA"] = lib.SymbolUnit_create_nanoampere,
        ["pA"] = lib.SymbolUnit_create_picoampere,
        ["Hz"] = lib.SymbolUnit_create_hertz,
        ["kHz"] = lib.SymbolUnit_create_kilohertz,
        ["MHz"] = lib.SymbolUnit_create_megahertz,
        ["GHz"] = lib.SymbolUnit_create_gigahertz,
        ["s"] = lib.SymbolUnit_create_second,
        ["ms"] = lib.SymbolUnit_create_millisecond,
        ["us"] = lib.SymbolUnit_create_microsecond,
        ["ns"] = lib.SymbolUnit_create_nanosecond,
        ["ps"] = lib.SymbolUnit_create_picosecond,
        ["m"] = lib.SymbolUnit_create_meter,
        ["mm"] = lib.SymbolUnit_create_millimeter,
        ["km"] = lib.SymbolUnit_create_kilometer,
        ["K"] = lib.SymbolUnit_create_kelvin,
        ["Ohm"] = lib.SymbolUnit_create_ohm,
        ["mOhm"] = lib.SymbolUnit_create_milliohm,
        ["kOhm"] = lib.SymbolUnit_create_kiloohm,
        ["MOhm"] = lib.SymbolUnit_create_megaohm,
        ["dimensionless"] = lib.SymbolUnit_create_dimensionless,
        ["%"] = lib.SymbolUnit_create_percent,
        ["rad"] = lib.SymbolUnit_create_radian,
    }
    -- Check for exact match
    if creators[unit_str] then return creators[unit_str]() end
    
    -- Fallback to JSON for complex units
    local s = lib.String_wrap(unit_str)
    local handle = lib.SymbolUnit_from_json_string(s)
    lib.String_destroy(s)
    return handle
end

song.register("SymbolUnit", {
    __mul = lib.SymbolUnit_multiplication,
    __div = lib.SymbolUnit_division,
    static_methods = {
        from_string = SymbolUnit.from_string,
        from_volt = lib.SymbolUnit_create_volt,
        from_millivolt = lib.SymbolUnit_create_millivolt,
        from_kilovolt = lib.SymbolUnit_create_kilovolt,
        from_ampere = lib.SymbolUnit_create_ampere,
        from_milliampere = lib.SymbolUnit_create_milliampere,
        from_microampere = lib.SymbolUnit_create_microampere,
        from_nanoampere = lib.SymbolUnit_create_nanoampere,
        from_picoampere = lib.SymbolUnit_create_picoampere,
        from_hertz = lib.SymbolUnit_create_hertz,
        from_kilohertz = lib.SymbolUnit_create_kilohertz,
        from_megahertz = lib.SymbolUnit_create_megahertz,
        from_gigahertz = lib.SymbolUnit_create_gigahertz,
        from_second = lib.SymbolUnit_create_second,
        from_millisecond = lib.SymbolUnit_create_millisecond,
        from_microsecond = lib.SymbolUnit_create_microsecond,
        from_nanosecond = lib.SymbolUnit_create_nanosecond,
        from_picosecond = lib.SymbolUnit_create_picosecond,
        from_meter = lib.SymbolUnit_create_meter,
        from_millimeter = lib.SymbolUnit_create_millimeter,
        from_kilometer = lib.SymbolUnit_create_kilometer,
        from_kelvin = lib.SymbolUnit_create_kelvin,
        from_ohm = lib.SymbolUnit_create_ohm,
        from_milliohm = lib.SymbolUnit_create_milliohm,
    },
    methods = {
        symbol = function(self)
            local h = lib.SymbolUnit_symbol(self)
            if h == nil then return "" end
            local ffi = require("ffi")
            local str = ffi.string(h.raw, h.length)
            lib.String_destroy(h)
            return str
        end,
        name = function(self)
            local h = lib.SymbolUnit_name(self)
            if h == nil then return "" end
            local ffi = require("ffi")
            local str = ffi.string(h.raw, h.length)
            lib.String_destroy(h)
            return str
        end,
        is_compatible_with = lib.SymbolUnit_is_compatible_with,
        convert_to = lib.SymbolUnit_convert_value_to,
    }
}, SymbolUnit)

return SymbolUnit

File: ocaml/src/physics/units/symbolunit.ml
open Ctypes
open Capi_bindings
open Error_handling

(* No opens needed - using qualified names *)

class type c_symbolunit_t = object
  method raw : unit ptr
end
class c_symbolunit (h : unit ptr) : c_symbolunit_t = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.symbolunit_destroy raw_val;
    Error_handling.raise_if_error ()
  ) self
end

module SymbolUnit = struct
  type t = c_symbolunit

  let copy (handle : t) : t =
    Error_handling.read handle (fun () ->
      let ptr = Capi_bindings.symbolunit_copy handle#raw in
      Error_handling.raise_if_error ();
      new c_symbolunit ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.symbolunit_from_json_string (Falcon_string.of_string json) in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let meter  : t =
    let ptr = Capi_bindings.symbolunit_create_meter () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let kilogram  : t =
    let ptr = Capi_bindings.symbolunit_create_kilogram () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let second  : t =
    let ptr = Capi_bindings.symbolunit_create_second () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let ampere  : t =
    let ptr = Capi_bindings.symbolunit_create_ampere () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let kelvin  : t =
    let ptr = Capi_bindings.symbolunit_create_kelvin () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let mole  : t =
    let ptr = Capi_bindings.symbolunit_create_mole () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let candela  : t =
    let ptr = Capi_bindings.symbolunit_create_candela () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let hertz  : t =
    let ptr = Capi_bindings.symbolunit_create_hertz () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let newton  : t =
    let ptr = Capi_bindings.symbolunit_create_newton () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let pascal  : t =
    let ptr = Capi_bindings.symbolunit_create_pascal () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let joule  : t =
    let ptr = Capi_bindings.symbolunit_create_joule () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let watt  : t =
    let ptr = Capi_bindings.symbolunit_create_watt () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let coulomb  : t =
    let ptr = Capi_bindings.symbolunit_create_coulomb () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let volt  : t =
    let ptr = Capi_bindings.symbolunit_create_volt () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let farad  : t =
    let ptr = Capi_bindings.symbolunit_create_farad () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let ohm  : t =
    let ptr = Capi_bindings.symbolunit_create_ohm () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let siemens  : t =
    let ptr = Capi_bindings.symbolunit_create_siemens () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let weber  : t =
    let ptr = Capi_bindings.symbolunit_create_weber () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let tesla  : t =
    let ptr = Capi_bindings.symbolunit_create_tesla () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let henry  : t =
    let ptr = Capi_bindings.symbolunit_create_henry () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let minute  : t =
    let ptr = Capi_bindings.symbolunit_create_minute () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let hour  : t =
    let ptr = Capi_bindings.symbolunit_create_hour () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let electronvolt  : t =
    let ptr = Capi_bindings.symbolunit_create_electronvolt () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let celsius  : t =
    let ptr = Capi_bindings.symbolunit_create_celsius () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let fahrenheit  : t =
    let ptr = Capi_bindings.symbolunit_create_fahrenheit () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let dimensionless  : t =
    let ptr = Capi_bindings.symbolunit_create_dimensionless () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let percent  : t =
    let ptr = Capi_bindings.symbolunit_create_percent () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let radian  : t =
    let ptr = Capi_bindings.symbolunit_create_radian () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let kilometer  : t =
    let ptr = Capi_bindings.symbolunit_create_kilometer () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let millimeter  : t =
    let ptr = Capi_bindings.symbolunit_create_millimeter () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let millivolt  : t =
    let ptr = Capi_bindings.symbolunit_create_millivolt () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let kilovolt  : t =
    let ptr = Capi_bindings.symbolunit_create_kilovolt () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let milliampere  : t =
    let ptr = Capi_bindings.symbolunit_create_milliampere () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let microampere  : t =
    let ptr = Capi_bindings.symbolunit_create_microampere () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let nanoampere  : t =
    let ptr = Capi_bindings.symbolunit_create_nanoampere () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let picoampere  : t =
    let ptr = Capi_bindings.symbolunit_create_picoampere () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let millisecond  : t =
    let ptr = Capi_bindings.symbolunit_create_millisecond () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let microsecond  : t =
    let ptr = Capi_bindings.symbolunit_create_microsecond () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let nanosecond  : t =
    let ptr = Capi_bindings.symbolunit_create_nanosecond () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let picosecond  : t =
    let ptr = Capi_bindings.symbolunit_create_picosecond () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let milliohm  : t =
    let ptr = Capi_bindings.symbolunit_create_milliohm () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let kiloohm  : t =
    let ptr = Capi_bindings.symbolunit_create_kiloohm () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let megaohm  : t =
    let ptr = Capi_bindings.symbolunit_create_megaohm () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let millihertz  : t =
    let ptr = Capi_bindings.symbolunit_create_millihertz () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let kilohertz  : t =
    let ptr = Capi_bindings.symbolunit_create_kilohertz () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let megahertz  : t =
    let ptr = Capi_bindings.symbolunit_create_megahertz () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let gigahertz  : t =
    let ptr = Capi_bindings.symbolunit_create_gigahertz () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let metersPerSecond  : t =
    let ptr = Capi_bindings.symbolunit_create_meters_per_second () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let metersPerSecondSquared  : t =
    let ptr = Capi_bindings.symbolunit_create_meters_per_second_squared () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let newtonMeter  : t =
    let ptr = Capi_bindings.symbolunit_create_newton_meter () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let newtonsPerMeter  : t =
    let ptr = Capi_bindings.symbolunit_create_newtons_per_meter () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let voltsPerMeter  : t =
    let ptr = Capi_bindings.symbolunit_create_volts_per_meter () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let voltsPerSecond  : t =
    let ptr = Capi_bindings.symbolunit_create_volts_per_second () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let amperesPerMeter  : t =
    let ptr = Capi_bindings.symbolunit_create_amperes_per_meter () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let voltsPerAmpere  : t =
    let ptr = Capi_bindings.symbolunit_create_volts_per_ampere () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let wattsPerMeterKelvin  : t =
    let ptr = Capi_bindings.symbolunit_create_watts_per_meter_kelvin () in
    Error_handling.raise_if_error ();
    new c_symbolunit ptr

  let equal (handle : t) (other : t) : bool =
    Error_handling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.symbolunit_equal handle#raw other#raw in
      Error_handling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    Error_handling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.symbolunit_not_equal handle#raw other#raw in
      Error_handling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.symbolunit_to_json_string handle#raw in
      Error_handling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let symbol (handle : t) : string =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.symbolunit_symbol handle#raw in
      Error_handling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let name (handle : t) : string =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.symbolunit_name handle#raw in
      Error_handling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let multiplication (handle : t) (other : t) : t =
    Error_handling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.symbolunit_multiplication handle#raw other#raw in
      Error_handling.raise_if_error ();
      new c_symbolunit result
    )

  let division (handle : t) (other : t) : t =
    Error_handling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.symbolunit_division handle#raw other#raw in
      Error_handling.raise_if_error ();
      new c_symbolunit result
    )

  let power (handle : t) (power : int) : t =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.symbolunit_power handle#raw power in
      Error_handling.raise_if_error ();
      new c_symbolunit result
    )

  let withPrefix (handle : t) (prefix : string) : t =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.symbolunit_with_prefix handle#raw (Falcon_string.of_string prefix) in
      Error_handling.raise_if_error ();
      new c_symbolunit result
    )

  let convertValueTo (handle : t) (value : float) (target : t) : float =
    Error_handling.multi_read [handle; target] (fun () ->
      let result = Capi_bindings.symbolunit_convert_value_to handle#raw value target#raw in
      Error_handling.raise_if_error ();
      result
    )

  let isCompatibleWith (handle : t) (other : t) : bool =
    Error_handling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.symbolunit_is_compatible_with handle#raw other#raw in
      Error_handling.raise_if_error ();
      result
    )

end